### DISCUSSION TITLE
El programa deja de funcionar al intentar procesar una textura animada

### DISCUSSION THREAD

**Discussion by LuishMC** [ComunidadAylas/PackSquash#114] [Q&A]
**Describir el error:**
Al intentar optimizar una textura con una gran resolución de altura, aparece un mensaje de error en la IU diciendo que la resolución de la altura es incorrecta y luego el programa se detiene, perdiendo todo el progreso de lo optimizado. Este error ocurre mayormente con las texturas de agua, lava y fuego en una resolución de anchura de 32x32 o más.

**Para reproducir:**
1. Descargar de internet cualquier paquete de recursos de alta resolución
2. Descomprimir el archivo y marcar el directorio de la carpeta en pack_directory en el archivo de configuración
3. Eliminar todos los archivos del paquete de recursos descomprimido excepto pack.mcmeta y las texturas animadas
4. Correr el programa

**Comment by AlexTMjugador** [comment]
Antes de nada, ¡gracias por usar PackSquash y hacernos llegar tus comentarios! :smile:

Por lo que describes en la incidencia, asumo que el error que comentas muestra un mensaje similar al siguiente:

```
! ruta/a/fichero.png: The texture width or height exceeds the configured maximum size. More information: <https://packsquash.page.link/Too-big-PNG-help>
```

Las versiones de PackSquash anteriores a la v0.3.1 pueden mostrar mensajes de error algo diferentes, pero haciendo referencia al tamaño de la imagen de todas formas.

Si, en efecto, este es el mensaje de error al que aludes, este comportamiento se trata de una decisión de diseño premeditada: por defecto, PackSquash impone un límite de resolución conservador de 8192 píxeles en cada dimensión para todas las texturas, ya que tamaños mayores pueden llegar a ser problemáticos.

Tales problemas son causados porque, internamente, el juego junta las texturas individuales de ciertos elementos, como bloques e ítems, en una textura más grande, para mejorar el rendimiento. No obstante, todas las GPU tienen un tamaño de textura máximo. En GPUs NVIDIA de la serie Turing y posterior, este límite es bastante alto, de 32768x32768 píxeles. Sin embargo, en GPUs más antiguas o de menores prestaciones, como las que se pueden encontrar en teléfonos móviles y tablets, este límite puede caer a, por ejemplo, 4096x4096 píxeles, y se puede exceder con facilidad por Minecraft cuando se están usando texturas de alta resolución. Cuando este límite se sobrepasa, la GPU rechaza crear la textura, y el paquete no se muestra correctamente.

En consecuencia, el comportamiento de PackSquash está pensado para que, como autor de un paquete de recursos, decidas si realmente quieres usar texturas de alta resolución, siendo consciente de que el paquete a lo mejor no funciona correctamente en dispositivos más limitados. Si estás seguro de que quieres usar texturas de alta resolución, basta con que incrementes el valor de la opción [`maximum_width_and_height`](https://github.com/ComunidadAylas/PackSquash/wiki/Options-files#maximum_width_and_height) a un valor que te parezca más razonable. Esta opción se puede aplicar individualmente por textura, así que es posible tener diferentes límites según sea conveniente.

Por ejemplo, para incrementar el límite de tamaño a 16384x16384 píxeles en todas las texturas puedes añadir las siguientes líneas a tu archivo de configuración:

```toml
['**/*.png']
maximum_width_and_height = 16384
```

¡Cuéntanos qué tal te funciona este cambio en la configuración!

**Comment by LuishMC** [comment]
> Antes de nada, ¡gracias por usar PackSquash y hacernos llegar tus comentarios! 😄
> 
> Por lo que describes en la incidencia, asumo que el error que comentas muestra un mensaje similar al siguiente:
> 
> ```
> ! ruta/a/fichero.png: The texture width or height exceeds the configured maximum size. More information: <https://packsquash.page.link/Too-big-PNG-help>
> ```
> 
> Las versiones de PackSquash anteriores a la v0.3.1 pueden mostrar mensajes de error algo diferentes, pero haciendo referencia al tamaño de la imagen de todas formas.
> 
> Si, en efecto, este es el mensaje de error al que aludes, este comportamiento se trata de una decisión de diseño premeditada: por defecto, PackSquash impone un límite de resolución conservador de 8192 píxeles en cada dimensión para todas las texturas, ya que tamaños mayores pueden llegar a ser problemáticos.
> 
> Tales problemas son causados porque, internamente, el juego junta las texturas individuales de ciertos elementos, como bloques e ítems, en una textura más grande, para mejorar el rendimiento. No obstante, todas las GPU tienen un tamaño de textura máximo. En GPUs NVIDIA de la serie Turing y posterior, este límite es bastante alto, de 32768x32768 píxeles. Sin embargo, en GPUs más antiguas o de menores prestaciones, como las que se pueden encontrar en teléfonos móviles y tablets, este límite puede caer a, por ejemplo, 4096x4096 píxeles, y se puede exceder con facilidad por Minecraft cuando se están usando texturas de alta resolución. Cuando este límite se sobrepasa, la GPU rechaza crear la textura, y el paquete no se muestra correctamente.
> 
> En consecuencia, el comportamiento de PackSquash está pensado para que, como autor de un paquete de recursos, decidas si realmente quieres usar texturas de alta resolución, siendo consciente de que el paquete a lo mejor no funciona correctamente en dispositivos más limitados. Si estás seguro de que quieres usar texturas de alta resolución, basta con que incrementes el valor de la opción [`maximum_width_and_height`](https://github.com/ComunidadAylas/PackSquash/wiki/Options-files#maximum_width_and_height) a un valor que te parezca más razonable. Esta opción se puede aplicar individualmente por textura, así que es posible tener diferentes límites según sea conveniente.
> 
> Por ejemplo, para incrementar el límite de tamaño a 16384x16384 píxeles en todas las texturas puedes añadir las siguientes líneas a tu archivo de configuración:
> 
> ```toml
> ['**/*.png']
> maximum_width_and_height = 16384
> ```
> 
> ¡Cuéntanos qué tal te funciona este cambio en la configuración!

Hola, muchas gracias por responder 😀. Pensé que tal vez no responderías, porque pensé que había escrito algo mal, pues es mi primera vez en GitHub XD. Había pensado en reportar varios errores en Discord, pero no tengo cuenta.

La verdad es que no había notado antes el enlace en el mensaje de error, pues apenas me daba tiempo de leerlo ya que enseguida el programa se cerraba y solamente podía leer "height" y otras cosas. Aunque no me molesta que la imágen no se procese, me molesta que se cierre y no pase por alto el archivo y copiándolo igual.

Edit: Ya me funciona correctamente

**Comment by AlexTMjugador** [comment]
> La verdad es que no había notado antes el enlace en el mensaje de error, pues apenas me daba tiempo de leerlo ya que enseguida el programa se cerraba y solamente podía leer "height" y otras cosas. Aunque no me molesta que la imágen no se procese, me molesta que se cierre y no pase por alto el archivo y copiándolo igual.

Para evitar que la ventana de PackSquash se cierre cuando termine, puedes ejecutarlo desde un terminal o, si usas Windows, símbolo del sistema. De esta forma podrás ver todos los mensajes que se muestren con facilidad.

Asumiendo que usas Windows, [hay páginas en español](https://es.digitaltrends.com/computadoras/como-usar-el-simbolo-del-sistema/) explicando cómo se abre el símbolo del sistema. Una vez en él, PackSquash se puede lanzar con el siguiente comando, sustituyendo las rutas como corresponda (es importante mantener las comillas dobles para evitar problemas con espacios):

```
"C:\ruta\a\ejecutable\de\packsquash.exe" "C:\ruta\a\tu\fichero\de\opciones.toml"
```

También es posible crear accesos directos o guiones (_scripts_) de procesamiento por lotes (normalmente conocidos por su extensión `.bat`) para ejecutar PackSquash de esta forma, pero eso es un poco más avanzado.

**Comment by LuishMC** [comment]
> Para evitar que la ventana de PackSquash se cierre cuando termine, puedes ejecutarlo desde un terminal o, si usas Windows, símbolo del sistema. De esta forma podrás ver todos los mensajes que se muestren con facilidad.
> 
> Asumiendo que usas Windows, [hay páginas en español](https://es.digitaltrends.com/computadoras/como-usar-el-simbolo-del-sistema/) explicando cómo se abre el símbolo del sistema. Una vez en él, PackSquash se puede lanzar con el siguiente comando, sustituyendo las rutas como corresponda (es importante mantener las comillas dobles para evitar problemas con espacios):
> 
> ```
> "C:\ruta\a\ejecutable\de\packsquash.exe" "C:\ruta\a\tu\fichero\de\opciones.toml"
> ```
> 
> También es posible crear accesos directos o guiones (_scripts_) de procesamiento por lotes (normalmente conocidos por su extensión `.bat`) para ejecutar PackSquash de esta forma, pero eso es un poco más avanzado.

Ya había probado usar PackSquash desde una terminal, pero actualmente lo uso por medio de un acceso directo ya que lo uso muy seguido. Pero aún así, el programa no sigue procesando el resto del paquete de texturas/recursos al dar error con la imágen, lo cual es lo que me molesta.

Por cierto, hace dos días intenté colocar un valor alto en maximum_width_and_height que me funcione con una textura que necesito y el programa no abre a menos que disminuya el valor. Específicamente no lo recuerdo, pero era en múltiplos del valor por defecto.

PD: Perdona por responder un poco tarde.

**Comment by AlexTMjugador** [comment]
> Por cierto, hace dos días intenté colocar un valor alto en maximum_width_and_height que me funcione con una textura que necesito y el programa no abre a menos que disminuya el valor. Específicamente no lo recuerdo, pero era en múltiplos del valor por defecto.
> 
> PD: Perdona por responder un poco tarde.

No te preocupes :smile:

El valor máximo para `maximum_width_and_height` es 65535, que a efectos prácticos desactiva el límite de tamaño. ¿Has comprobado que sea 65535 o menos?

**Comment by LuishMC** [comment]
> El valor máximo para `maximum_width_and_height` es 65535, que a efectos prácticos desactiva el límite de tamaño. ¿Has comprobado que sea 65535 o menos?

No me funciona, al principio intenté con 8192. Por lo visto, el límite en mi PC es alrededor de 6144

**Comment by AlexTMjugador** [comment]
Es un poco extraño que el límite sea ese. Ver los mensajes de error exactos que PackSquash muestra ayudaría bastante a diagnosticar el problema.

**Comment by LuishMC** [comment]
> Es un poco extraño que el límite sea ese. Ver los mensajes de error exactos que PackSquash muestra ayudaría bastante a diagnosticar el problema.

No muestra ningún mensaje cuando tengo un valor así de alto, solamente la consola se queda en blanco después de iniciar y luego se cierra

**Comment by AlexTMjugador** [comment]
He movido la incidencia a una discusión ("post") ya que se estaba desvirtuando un poco su tema original, pero podemos seguir hablando por la discusión.

> No muestra ningún mensaje cuando tengo un valor así de alto, solamente la consola se queda en blanco después de iniciar y luego se cierra

Ello podría deberse a varias causas. ¿Cuánta memoria principal (RAM) tiene disponible tu equipo cuando ejecutas PackSquash? Si tu PC tiene muy poca memoria disponible, puede pasar que el procesado de imágenes de mayor tamaño consuma más memoria de la disponible y no se pueda completar correctamente.

**Reply by LuishMC** [reply]
Tengo 4 GB de RAM y un CPU Intel i3. Ya sé, es muy malo, mi procesador original se dañó y lo tuve que reemplazar por este. xd

**Reply by AlexTMjugador** [reply]
Puedes probar a incrementar el tamaño del archivo de paginación (hay instrucciones [aquí](https://answers.microsoft.com/es-es/windows/forum/all/c%C3%B3mo-incrementar-la-memoria-virtual-en/c877e207-e5e8-4cd0-9f4b-9d643b9b269d?auth=1)) para usar tu memoria secundaria (disco duro, SSD) como si fuese RAM. Si la causa de los problemas es disponer de insuficiente memoria, eso puede arreglarlo, aunque el rendimiento será bastante malo, ya que la memoria secundaria es mucho más lenta que la RAM.

Desarrollo PackSquash en un PC con 16 GiB de RAM, así que añadir la memoria virtual necesaria para que la suma de RAM instalada + archivo de paginación sea de 16 GiB es una buena idea. Si no tienes el espacio libre necesario para llegar a este objetivo, puedes probar con totales de 8 y 12 GiB.

**Reply by LuishMC** [reply]
Que tal si abro un nuevo problema marcado como sugerencia que tal vez pasen por alto este problema y el #101?

**Reply by AlexTMjugador** [reply]
Me parece que de momento es mejor no crear una nueva incidencia, dado que de momento no tenemos muy claro cómo reproducir el problema.

¿Hubo algún avance al respecto? :smile:

**Reply by LuishMC** [reply]
Bueno, aquí tengo dos sugerencias. La primera pasaría por alto este problema y el #101 y la segunda haría ambos problemas menos molestos:

1. Aprovechando que el programa usa una consola como IU, tal vez hacer que cuando de un error de archivo te de opciones de qué hacer con él. He visto programas de consolas que te preguntan qué quieres hacer a los que se les debe responder presionando una tecla o escribiendo algo y luego presionando Enter, y si se hace algo parecido? Por ejemplo, marcar el número 1 cerraría el programa, marcar el 2 eliminaría el archivo y seguiría procesando el resto, y marcar el 3 copiaría el archivo igual al paquete y continuaría procesando.
2. Añadir a los archivos de configuración una opción para que se escriba un archivo de texto donde se documente todo lo ocurrido, desde la última inicialización del programa hasta que se cerró por última vez. Esto me ayudaría mucho ya que el programa cierra inmediatamente al dar error y siempre me cuesta identificar el archivo que falla.

**Reply by AlexTMjugador** [reply]
El segundo punto ya se puede hacer si en el comando de lanzar PackSquash se indica que el flujo de salida de errores se debe redireccionar a un archivo:

```
C:\ruta\a\packsquash.exe "C:\ruta\a\fichero\de\opciones.toml" 2>"C:\ruta\a\fichero\de\errores.txt"
```

Sobre el primer punto, me parece que es una buena idea. Habría que tener un poco de cuidado con detectar si PackSquash está siendo usado de forma automatizada o no, y quizás también añadir opciones para recordar la opción seleccionada para el resto de archivos, ya que si no puede pasar que el usuario tenga que estar marcando 3 miles de veces, pero por lo demás concuerdo en que puede mejorar la experiencia. ¡Gracias!

En cuanto a #101, me parece que lo ideal sería que tales archivos de audio fuesen aceptados por PackSquash, ya que si bien tienen un formato algo extraño, es técnicamente válido.
